cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/ProvisioningService.java

```diff
@@ -521,20 +523,28 @@ public void initializeProvisionersAndExecutors() {
     contextExecutor.allowCoreThreadTimeOut(true);
     this.contextExecutor = contextExecutor;
 
-    Map<String, Provisioner> provisioners = provisionerProvider.loadProvisioners();
+    Map<String, Provisioner> provisioners = new HashMap<>(provisionerProvider.loadProvisioners());
     Map<String, ProvisionerConfig> provisionerConfigs =
       provisionerConfigProvider.loadProvisionerConfigs(provisioners.values());
     LOG.debug("Provisioners = {}", provisioners);
     Map<String, ProvisionerDetail> details = new HashMap<>(provisioners.size());
-    for (Map.Entry<String, Provisioner> provisionerEntry : provisioners.entrySet()) {
+    boolean tetheringEnabled = cConf.getBoolean(Constants.Tethering.TETHERING_SERVER_ENABLED);
+    Iterator<Map.Entry<String, Provisioner>> iterator = provisioners.entrySet().iterator();
+    while (iterator.hasNext()) {
+      Map.Entry<String, Provisioner> provisionerEntry = iterator.next();
       String provisionerName = provisionerEntry.getKey();
+      // Ignore the tethering provisioner if tethering is disabled
+      if (!tetheringEnabled && provisionerName.equals(TetheringProvisioner.TETHERING_NAME)) {
+        iterator.remove();
```

Comment:
it's better not to use iterator.remove(). A decent number of implementations don't actually implement it, and some others will also have issues if you remove while iterating.

Instead, it seems simpler to do a provisioners.remove(TetheringProvisioner.TETHERING_NAME) if tethering is disabled.

Reply:
done